src/Adapter/AbstractObjectModelRepository.php

```diff
@@ -66,15 +72,30 @@ protected function assertObjectModelExists(int $id, string $objectTableName, str
      * @param int $id
      * @param string $objectModelClass
      * @param string $exceptionClass
+     * @param ShopId|null $shopId
      *
      * @return ObjectModel
      *
      * @throws CoreException
      */
-    protected function getObjectModel(int $id, string $objectModelClass, string $exceptionClass): ObjectModel
+    protected function getObjectModel(int $id, string $objectModelClass, string $exceptionClass, ?ShopId $shopId = null): ObjectModel
     {
         try {
-            $objectModel = new $objectModelClass($id);
+            if (null !== $shopId) {
+                $this->checkShopAssociation($id, $objectModelClass, $shopId);
+            }
+
+            // Special case for Product because it has an additional full parameter
+            if ($objectModelClass === Product::class) {
+                $objectModel = new Product($id, false, null, $shopId ? $shopId->getValue() : null);
+            } else {
+                $objectModel = new $objectModelClass($id, null, $shopId ? $shopId->getValue() : null);
+            }
```

Comment:
You can extract this into a `constructObject(int $id, ShopId $shopId): ObjectModel` method, then leave it to the concrete class to override it for Product

Reply:
Ok I got it!! 😅